### ISSUE TITLE
Error when instantiating a class with an optional, type-hinted interface as parameter

### ISSUE THREAD

**Issue by masadelante** [rdlowrey/auryn#144]
In PHP 5.6, when trying to instantiate a class with the following constructor containing variadic arguments:

`public function __construct(Configuration $config, LdapConnectionInterface ...$connections)`

Auryn throws an exception because it tries to instantiate an interface:

`Injection definition required for non-concrete parameter
$interface of type
LdapTools\Connection\LdapConnectionInterface`

The error occurs because the injector doesn't check whether the parameter is variadic, and therefore optional, before trying to instantiate a parameter from the type hint. I'd suggest inserting an additional elseif before the last elseif statement in the method below, to check if the parameter is variadic.

It's necessary to continue to the next iteration of the foreach loop, since variadic arguments can also be non-existent, and simply returning a **null** value leads to errors.

```
private function provisionFuncArgs(\ReflectionFunctionAbstract $reflFunc, array $definition, array $reflParams = null) {
    $args = array();

    // @TODO store this in ReflectionStorage
    if (!isset($reflParams)) {
        $reflParams = $reflFunc->getParameters();
    }

    foreach ($reflParams as $i => $reflParam) {
        $name = $reflParam->name;

        if (isset($definition[$i]) || array_key_exists($i, $definition)) {
            // indexed arguments take precedence over named parameters
            $arg = $definition[$i];
        } elseif (isset($definition[$name]) || array_key_exists($name, $definition)) {
            // interpret the param as a class name to be instantiated
            $arg = $this->make($definition[$name]);
        } elseif (($prefix = self::A_RAW . $name) && (isset($definition[$prefix]) || array_key_exists($prefix, $definition))) {
            // interpret the param as a raw value to be injected
            $arg = $definition[$prefix];
        } elseif (($prefix = self::A_DELEGATE . $name) && isset($definition[$prefix])) {
            // interpret the param as an invokable delegate
            $arg = $this->buildArgFromDelegate($name, $definition[$prefix]);
        } elseif (($prefix = self::A_DEFINE . $name) && isset($definition[$prefix])) {
            // interpret the param as a class definition
            $arg = $this->buildArgFromParamDefineArr($definition[$prefix]);
        } elseif (!$arg = $this->buildArgFromTypeHint($reflFunc, $reflParam)) {
            $arg = $this->buildArgFromReflParam($reflParam);
        }

        $args[] = $arg;
    }

    return $args;
}
```

Like this:

```
        } elseif( method_exists($reflParam, 'isVariadic') && $reflParam->isVariadic() ) {
            continue;
        } elseif (!$arg = $this->buildArgFromTypeHint($reflFunc, $reflParam)) {
            $arg = $this->buildArgFromReflParam($reflParam);
        }
```

**Comment by J7mbo** [comment]
Is this perhaps related to #134?

**Comment by masadelante** [comment]
It is, though not entirely. Issue #134 and the proposed workaround of using delegates do not deal fully with the scenario where the variadic arguments do not contain anything at all.

I agree that it's useful to delegate object instantiation to a separate class when a variable number of dependencies need to be injected.

However, when the number of variadic arguments is zero, it would be more elegant to just treat them like any other optional argument in the injector, i.e. ignore them.

**Comment by kelunik** [comment]
Closed via #152.

**Comment by kelunik** [comment]
Released v1.4.2.
